File: 1755.closest-subsequence-sum.cs
/*
 * @lc app=leetcode id=1755 lang=csharp
 *
 * [1755] Closest Subsequence Sum
 */

// @lc code=start
using System;
using System.Collections.Generic;

public class Solution {
    private List<int> dpSumBitMaskSorted(List<int> nums) {
        int l = nums.Count;
        List<int> result = new List<int> { 0 };
        for (int i = 0; i < l; i++) {
            for (int j = (int)Math.Pow(2, i); j < Math.Pow(2, i + 1); j++) {
                result.Add(nums[i] + result[j - (int)Math.Pow(2, i)]);
            }
        }
        result.Sort();
        return result;
    }

    private static int bisectLeft<T>(IList<T> arr, T x, int lo=0, int hi=-1) where T : IConvertible, IComparable, IComparable<T>, IEquatable<T>
    {
        hi = (hi == -1) ? arr.Count : hi;
        while (lo < hi)
        {
            int mid = lo + (hi - lo) / 2;
            if (arr[mid].CompareTo(x) < 0)
            {
                lo = mid + 1;
            }
            else
            {
                hi = mid;
            }
        }
        return lo;     
    }

    public int MinAbsDifference(int[] nums, int goal) {
        int n = nums.Length;
        if (n == 1) {
            return Math.Min(Math.Abs(nums[0] - goal), Math.Abs(goal));
        } else {
            List<int> left = new List<int>();
            List<int> right = new List<int>();
            for (int i = 0; i < n / 2; i++) {
                left.Add(nums[i]);
            }
            for (int i = n / 2; i < n; i++) {
                right.Add(nums[i]);
            }
            List<int> leftSumSorted = dpSumBitMaskSorted(left);
            List<int> rightSumSorted = dpSumBitMaskSorted(right);
            int nr = rightSumSorted.Count;
            int result = int.MaxValue;
            foreach (int i in leftSumSorted) {
                int ind = bisectLeft<int>(rightSumSorted, goal - i);
                if (ind - 1 >= 0 && ind - 1 < nr) {
                    result = Math.Min(result, Math.Abs(rightSumSorted[ind - 1] + i - goal));
                }
                if (ind >= 0 && ind < nr) {
                    result = Math.Min(result, Math.Abs(rightSumSorted[ind] + i - goal));
                }
                if (ind + 1 >= 0 && ind + 1 < nr) {
                    result = Math.Min(result, Math.Abs(rightSumSorted[ind + 1] + i - goal));
                }
            }
            return result;
        }
    }
}

// @lc code=end

File: 1755.closest-subsequence-sum.cpp
/*
 * @lc app=leetcode id=1755 lang=cpp
 *
 * [1755] Closest Subsequence Sum
 */

// @lc code=start
#include<vector>
#include<algorithm>
using std::abs;
using std::lower_bound;
using std::make_move_iterator;
using std::min;
using std::sort;
using std::vector;
class Solution {
private:
    vector<int> dpSumBitMaskSorted(vector<int>& nums) {
    int l = nums.size();
    vector<int> result { 0 };
    for (int i = 0; i < l; i++) {
        for (int j = (1 << i); j < (1 << (i + 1)); j++) {
            result.push_back(nums[i] + result[j - (1 << i)]);
        }
    }
    sort(result.begin(), result.end());
    return result;
}

public:
    int minAbsDifference(vector<int>& nums, int goal) {
        int n = nums.size();
        if (n == 1) {
            return min(abs(nums[0] - goal), abs(goal));
        } else {
            vector<int> right(make_move_iterator(nums.begin() + nums.size()/2), make_move_iterator(nums.end()));
            nums.erase(nums.begin() + nums.size()/2, nums.end());

            vector<int> leftSumSorted = dpSumBitMaskSorted(nums);
            vector<int> rightSumSorted = dpSumBitMaskSorted(right);
            int nr = rightSumSorted.size();
            int result = INT_MAX;
            for (int i : leftSumSorted) {
                int ind = lower_bound(rightSumSorted.begin(), rightSumSorted.end(), goal - i) - rightSumSorted.begin();
                if (ind - 1 >= 0 && ind - 1 < nr) {
                    result = min(result, abs(rightSumSorted[ind - 1] + i - goal));
                }
                if (ind >= 0 && ind < nr) {
                    result = min(result, abs(rightSumSorted[ind] + i - goal));
                }
                if (ind + 1 >= 0 && ind + 1 < nr) {
                    result = min(result, abs(rightSumSorted[ind + 1] + i - goal));
                }
            }
            return result;
        }        
    }
};
// @lc code=end

